### PR TITLE
Implement Decoder::Take to selectively read a chunk with indices

### PR DESCRIPTION
**Pull request by eddyxu**
* Add `Decoder::Take(arrow::Array indices)` interface
* Add STL adaptor to convert a vector / initialize_list to array
* Add tests.

### CODE REVIEW

File: cpp/src/lance/arrow/stl.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#include <initializer_list>
+#include <memory>
+#include <vector>
+
+namespace lance::arrow {
+
+/// \file STL adaptors between Arrow and STL.
+
+/// Convert a vector to Apache Arrow Array.
+template <typename T>
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/stl.h:29]: Nice, no more giant cascade

File: cpp/src/lance/encodings/plain.h [hunk 2]
```diff
@@ -68,6 +68,28 @@ class PlainDecoder : public Decoder {
     return std::make_shared<ArrayType>(length.value(), buf);
   }
 
+  ::arrow::Result<std::shared_ptr<::arrow::Array>> Take(
```
Comment (changhiskhan) [inline cpp/src/lance/encodings/plain.h:71]: Dumb question: if we want say OFFSET 1000 LIMIT 10000, is this the method that gets called? If yes, does this mean the `indices` have to contain all of range(1000,10000)?

**Review by changhiskhan — APPROVED** [review]
LGTM with a couple of minor nits

File: cpp/src/lance/encodings/plain.h (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/encodings/plain.h:71]: There will be two ways to "get an array from decoder", which will be implemented in #32 eventually.

So first, is scan without filter, that, will just call `Decoder::ToArray(start, length)` for performance.

Bu with filters, the logic will essentially be:

```c++
Array mask;
for (auto filtered_column :  all_filter_columns) {
   mask &= filtered_column;  // or `mask |= filtered_column` depends on the expression
}
auto indices = MaskToIndices(mask);
for (auto projected_field : all_projected_field) {
   project_field->GetDecoder()->Take(indices)
}
```

Comment (eddyxu) [reply cpp/src/lance/encodings/plain.h:71]: Oh, also, the `indices` are within a "chunk" / "page".

**Pull request merged**